src/System.IO.FileSystem/src/System/IO/Win32FileStream.cs

```diff
@@ -1291,19 +1305,45 @@ private Task WriteInternalAsync(byte[] array, int offset, int numBytes, Cancella
                 _readLen = 0;
             }
 
-            int n = _bufferSize - _writePos;
-            if (numBytes <= n)
+            int remainingBuffer = _bufferSize - _writePos;
+
+            // if the incoming write fits in the remaining buffer, treat
+            // as synchronous since memcopy is so fast
+            if (numBytes <= remainingBuffer)
             {
                 if (_writePos == 0) _buffer = new byte[_bufferSize];
                 Buffer.BlockCopy(array, offset, _buffer, _writePos, numBytes);
                 _writePos += numBytes;
-                
+
                 return Task.CompletedTask;
             }
 
-            if (_writePos > 0)
-                FlushWrite(false);
+            // if the incoming write fits in the current buffer + next buffer,
+            // only perform a single write to flush the existing buffer and fill
+            // the next buffer with the remainder
+            if (numBytes <= (_bufferSize + remainingBuffer))
+            {
+                if (_writePos == 0) _buffer = new byte[_bufferSize];
+
+                Buffer.BlockCopy(array, offset, _buffer, _writePos, remainingBuffer);
+
+                byte[] oldBuffer = _buffer;
+                _buffer = new byte[_bufferSize];
```

Comment:
So does this mean that every time we hit this path of going past the end of the current buffer, we're going to allocate a new one?  That seems quite costly.

This would seem to put us between and a rock and a hard place.  The caller of WriteAsync needs to be able to use the byte[] they provide when the call returns, so we can't just asynchronously flush the buffer and then asynchronously write the data they provide, because as soon as we start the asynchronous flush, we'll return to them, but an asynchronous operation using their buffer will still be active.  An alternative is what the code already employs: do the flush synchronously, but that then mostly negates the value of using WriteAsync in the first place (any writes of a size less than the buffer will end up either synchrously writing to the buffer or synchronously flushing), and that's what this PR is trying to address. Ugh.

I wonder if some small buffer pool is warranted, e.g. we allow only up to N buffer allocations, which are pooled, and if we try to get a buffer from the pool and we've already allocated N, then we fall back to synchronous flushing.

Or maybe in the case of overflowing the buffer, we support growing the buffer used to be the size of what's already cached plus the size of whatever the developer supplied (we could still do flushes when we hit the original _bufferSize, but we'd have enough room to store all of the user supplied data).  In degenerate situations, the buffer could get extremely large, but maybe for such cases we set of a threshold and do synchronous flushes beyond that.

@ericstj, you've probably thought about this before...?